### ISSUE TITLE
Expose access to sub-contract calls, their args, their results, their events via meta or system events or debug events

### ISSUE THREAD

**Issue by leighmcculloch** [stellar/rs-soroban-env#587]
An observer of the network should be able to see the sub-contract calls to other contracts and those calls args, results, events, etc. This is relatively important for auditing, and for block explorer tools to be able to provide complete transparency into what is going on during invocations.

An example of tools in other ecosystems that display this type of information:
- https://etherscan.io/tx/0x7dec0cefd3306112871d3bba656b9f0f6ba336290f1f0558f8262323b6d730e5/advanced#internal

Maybe this information can be included in tx meta, or system events, or debug events.

Whatever location is chosen, it's relatively important this information is accessible in a structured format so that tools and systems can ingest it.

Discussion: https://discord.com/channels/897514728459468821/1046843869926527017

**Comment by MonsieurNicolas** [comment]
Yes: I actually discussed this with @graydon a few weeks ago as this allows to really cleanup where error codes are needed vs debug information for developers.

I think this creates an opportunity for rich feedback for developers (akin to the Javascript "console").

The TLDR; goes something like this:
* add a mechanism to optionally emit more data into the "debug" log stream
    * this is not included in the system and contract events as we do not want this to be hashed
    * it could be a separate event category to make it easier to filter and avoid mixing with user logs (`log_value` type of stuff), we would still return a single vector that represents events in execution order.
* we can probably have multiple levels of diagnostics
   * sub contract calls/returns (debug level)
       *  arguments on invoke (trace level)
       * result on return (trace level)
    * host function invocation (trace level)
* the kind of diagnostic/debug data emitted is environment specific
    * this may have significant impact on the instance and would probably be disabled by default on internet facing services (metering would also not take this into account) *but* it's plausible that it would be enabled in a reduced capacity for select transactions
* in `preflight`
   * those events are returned as a new vector of events `diagnostics_events[]` alongside other data like `footprint`
* in CLI
   * just pretty print to the console
* in `core`
   * core can be run with a global setting similar to log level that allows to adjust the default diagnostic level
   * diagnostic level for specific transactions can be specified when submitting transactions (new argument to the `tx` endpoint), which will override the default diagnostic level for that transaction when it gets applied
   * when a transaction gets _applied_ (during ledger close)
       * a new `diagnosticsEvents[]` gets populated in the meta in its non-hashed portion, something like:
```
struct TransactionMetaV3
{
    LedgerEntryChanges txChangesBefore; // tx level changes before operations
                                        // are applied if any
    OperationMeta operations<>;         // meta for each operation
    LedgerEntryChanges txChangesAfter;  // tx level changes after operations are
                                        // applied if any
    ContractEvent events<>;            // custom events populated by the
                                        // contracts themselves
    TransactionResult txResult;

    Hash hashes[3];                     // stores sha256(txChangesBefore, operations, txChangesAfter),
                                        // sha256(events), and sha256(txResult)

    DiagnosticEvents diagnosticsEvents<>;          // optional diagnostics events
};

```

**Comment by MonsieurNicolas** [comment]
Something that came up is around how to ensure that `diagnosticsEvents` is not used to attack instances (core nodes or downstream systems that use that data).

As those are not part of the core protocol (ie: there is no consensus on them), I think we are pretty flexible on policies that govern how they're generated.

I'd like to start with the simplest policy possible (and we can always add more later), maybe one of those would work (`N` is configured by the node operator):
* `diagnosticEvents` are stored in a circular buffer of sorts (so contain the last `N` diagnostic events) - I think this was the original idea that @graydon wanted to implements
* `diagnosticEvents` only contain the first `N` events
* `N` could also be not a number of bytes instead of a number of events

**Comment by orbitlens** [comment]
Not sure about the "circular buffer" approach, this will be useless for us if the number of generated events exceeds `N`, at least for purpose of building a subcontracts calls tree.

The option _"`diagnosticEvents` only contain the first N events"_ looks ideal in this regard, since it makes the output more predictable.

**Comment by MonsieurNicolas** [comment]
Yeah first N or last N are useful in different situations so maybe we just need both.

As this is not part of the protocol, this can be configured differently depending on use cases/environments.
The only reason to have a limit in the first place is to avoid getting DoSed in some situations (especially important if this is something exposed by preflight on internet facing endpoints), in your case I imagine that you'd have a dedicated private endpoint so the concern is more protecting against a rogue contract that produces too much diagnostic data. In any case, to be continued as we work through this.

**Comment by sisuresh** [comment]
Something that came up recently - [These](https://github.com/stellar/rs-soroban-env/pull/640) diagnostic events work just like normal contract events wrt failures. If a contract call fails, those events will be rolled back and not reported. This sounds like the desired behavior for downstream consumers like stellar-expert. Do we want these diagnostic events to be returned on failure as well? I'm thinking no ([here's](https://github.com/stellar/stellar-core/pull/3587#discussion_r1107916888) a comment I made explaining some weird scenarios with this) but I can see the argument for including them to debug failures as well so wanted to get feedback.

**Comment by sisuresh** [comment]
Closed by https://github.com/stellar/rs-soroban-env/pull/640
